les/vflux/server/prioritypool.go

```diff
@@ -500,3 +515,130 @@ func (pp *PriorityPool) updatePriority(node *enode.Node) {
 	}
 	updates = pp.tryActivate()
 }
+
+// CapacityCurve is a snapshot of the priority pool contents in a format that can efficiently
+// estimate how much capacity could be granted to a given node at a given priority level.
+type CapacityCurve struct {
+	points       []curvePoint       // curve points sorted in descending order of priority
+	index        map[enode.ID][]int // curve point indexes belonging to each node
+	exclude      []int              // curve point indexes of excluded node
+	excludeFirst bool               // true if activeCount == maxCount
+}
+
+type curvePoint struct {
+	freeCap, freeCount uint64 // available capacity and node count at the current priority level
+	nextPri            int64  // next priority level where more capacity will be available
+}
+
+// GetCapacityCurve returns a new or recently cached CapacityCurve based on the contents of the pool
+func (pp *PriorityPool) GetCapacityCurve() *CapacityCurve {
+	pp.lock.Lock()
+	defer pp.lock.Unlock()
+
+	now := pp.clock.Now()
+	dt := time.Duration(now - pp.ccUpdatedAt)
+	if !pp.ccUpdateForced && pp.cachedCurve != nil && dt < time.Second*10 {
+		return pp.cachedCurve
+	}
+
+	pp.ccUpdateForced = false
+	pp.ccUpdatedAt = now
+	curve := &CapacityCurve{
+		index: make(map[enode.ID][]int),
+	}
+	pp.cachedCurve = curve
+
+	var excludeID enode.ID
+	excludeFirst := pp.maxCount == pp.activeCount
+	for pp.activeCap > 0 {
+		cp := curvePoint{
+			freeCap: pp.maxCap - pp.activeCap,
+		}
+		tempCap := cp.freeCap + 1
+		pp.activeCap += tempCap
+		var next *ppNodeInfo
+		next, cp.nextPri = pp.enforceLimits()
+		id := next.node.ID()
+		if excludeFirst {
+			curve.excludeFirst = true
+			excludeID = id
+			excludeFirst = false
+		}
+		curve.index[id] = append(curve.index[id], len(curve.points))
+		curve.points = append(curve.points, cp)
+		pp.activeCap -= tempCap
+	}
+	pp.finalizeChanges(false)
+	curve.points = append(curve.points, curvePoint{
+		freeCap: pp.maxCap,
+		nextPri: math.MaxInt64,
+	})
+	if curve.excludeFirst {
+		curve.exclude = curve.index[excludeID]
+	}
+	return curve
+}
+
+// Exclude returns a CapacityCurve with the given node excluded from the original curve
+func (cc *CapacityCurve) Exclude(id enode.ID) *CapacityCurve {
+	if exclude, ok := cc.index[id]; ok {
+		return &CapacityCurve{
+			points:  cc.points,
+			index:   cc.index,
+			exclude: exclude,
```

Comment:
The excludeFirst shouldn't be copied?

Reply:
No, and neither is `exclude` copied. If the first node was excluded, that's because the total node count was reached. If another node needs to be excuded anyways (because the CQ requesting node is already connected) then that exclusion will already ensure that the total node count is not exceeded (and therefore the first node does not need to be dropped anymore).